watertap/unit_models/crystallizer.py

```diff
@@ -0,0 +1,696 @@
+            #  mass_frac_after_evap = SOLIDS IN + LIQUID IN - VAPOUR OUT
+            mass_frac_after_evap = b.properties_in[0].flow_mass_phase_comp['Liq', j] \
+            / (sum(b.properties_in[0].flow_mass_phase_comp['Liq', j] for j in solute_set)
+                + b.properties_in[0].flow_mass_phase_comp['Liq', 'H2O']
+                - b.properties_vapor[0].flow_mass_phase_comp['Vap', 'H2O']
+                )
+            # return (b.relative_supersaturation[j] * b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j] == 
+            # (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+            # )
+            return (b.relative_supersaturation[j] == 
+                (mass_frac_after_evap - b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j])
+                / b.properties_out[0].solubility_mass_frac_phase_comp['Liq', j]
+                )
+
+
+
+
+
+
+        # 4. Fix flows of empty solid, liquid and vapour streams
+        # (i) Fix solids: liquid and vapour flows must be zero
+        @self.Constraint(self.properties_solids[0].phase_component_set,
+            doc="Empty components in solids stream")
+        def eq_solids_composition(b, p, j):
+            if p != 'Sol':
+                return (b.properties_solids[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+        # (ii) Fix liquids: solid and vapour flows must be zero
+        @self.Constraint(self.properties_out[0].phase_component_set,
+            doc="Empty components in liquid stream")
+        def eq_liquid_composition(b, p, j):
+            if p != 'Liq':
+                return (b.properties_out[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+        # (iii) Fix vapor: solid and vapour flows must be zero. Vapour is also pure solvent.
+        @self.Constraint(self.properties_vapor[0].phase_component_set,
+            doc="Empty components in vapour stream")
+        def eq_vapor_composition(b, p, j):
+            if p != 'Vap':
+                return (b.properties_vapor[0].flow_mass_phase_comp[p, j] == 1e-8 * pyunits.kg/pyunits.s)
+            else:
+                return Constraint.Skip
+
+
+        # 5. Add an energy balance for the system
+        ## (iii) Enthalpy balance: based on Lewis et al. Enthalpy is exothermic, and the hC in property package is -ve
+        @self.Constraint(doc="Enthalpy balance over crystallization system")
+        def eq_enthalpy_balance(b):
+            return (b.properties_in[0].enth_flow 
+                - b.properties_out[0].enth_flow - b.properties_vapor[0].enth_flow - b.properties_solids[0].enth_flow
+                + self.work_mechanical[0]
+                - (b.properties_solids[0].flow_vol_phase['Sol'] * b.properties_solids[0].dens_mass_solute['Sol'] * b.properties_solids[0].dh_crystallization['NaCl'])
```

Comment:
Can't you grab the mass flowrate directly `b.properties_solids[0].flow_mass_phase_comp['Sol', j]? You could also sum across the solutes, whenever we do multiple solutes.

Reply:
Done.